### PR TITLE
[TorchToLinalg] : Lower count_nonzero.dim_IntList to Linalg-on-Tensors

### PR DESCRIPTION
**Pull request by um-urosmarkovic**
- Lowered count_nonzero.dim_IntList op to Linalg-on-Tensors
- Decomposed aten.count_nonzero using aten.ne.Scalar and aten.sum.dim_IntList
- Added tests to torch-mlir/projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py

### CODE REVIEW

**Comment by um-urosmarkovic** [comment]
Hi @vivekkhandelwal1 , could you please approve the workflows when you get the chance? Thanks!

**Comment by um-urosmarkovic** [comment]
Hi @vivekkhandelwal1 , I’ve rebased the branch and the CI workflows need maintainer approval again. Could you please click “Approve and run” when you have a moment? Thanks!

**Comment by vivekkhandelwal1** [comment]
> Hi @vivekkhandelwal1 , I’ve rebased the branch and the CI workflows need maintainer approval again. Could you please click “Approve and run” when you have a moment? Thanks!

Hi @um-urosmarkovic, there's an issue with the CI, hence no jobs are running. It has to be re-run once that is fixed.

**Comment by um-urosmarkovic** [comment]
Hi @vivekkhandelwal1, the CI workflows have passed. When you get a chance, could you please review? Thanks in advance!

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 1]
```diff
@@ -2496,3 +2496,54 @@ def TraceUnsignedIntModule_basic(module, tu: TestUtils):
...
+
+    @export
+    @annotate_args([None, ([2, 3, 4], torch.float32, True)])
+    def forward(self, x):
+        return torch.ops.aten.count_nonzero(x, self.dim)
+
+
+@register_test_case(module_factory=lambda: CountNonzeroDimIntListModuleF32())
+def CountNonzeroDimIntListModuleF32_basic(module, tu: TestUtils):
+    module.forward(tu.rand(2, 3, 4).to(torch.float32))
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py:2517]: You don't need to do `to(torch.float32)` since it by default generates a tensor with `dtype=torch.float32`.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 2]
```diff
@@ -2496,3 +2496,54 @@ def TraceUnsignedIntModule_basic(module, tu: TestUtils):
...
+
+    @export
+    @annotate_args([None, ([2, 3, 4], torch.int64, True)])
+    def forward(self, x):
+        return torch.ops.aten.count_nonzero(x, self.dim)
+
+
+@register_test_case(module_factory=lambda: CountNonzeroDimIntListModuleI64())
+def CountNonzeroDimIntListModuleI64_basic(module, tu: TestUtils):
+    module.forward(tu.randint(2, 3, 4).to(torch.int64))
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py:2533]: Same as above.

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py [hunk 3]
```diff
@@ -1436,6 +1436,23 @@ def aten〇rot90〡shape(self: List[int], k: int = 1, dims: List[int] = (0, 1,))
...
+    Invocation(TensorOfShape(2, 3, 4), dim=[2]), # Maximum valid dim.
+    ErrorInvocation(TensorOfShape(2, 3, 4), dim=[-4]), # Test dim out of bound.
+     ErrorInvocation(TensorOfShape(2, 3, 4), dim=[1,-4]), # Test dim out of bound.
+    ErrorInvocation(TensorOfShape(2, 3, 4), dim=[3]), # Test dim out of bound.
+])
+def aten〇count_nonzero〇dim_IntList〡shape(self: List[int], dim: List[int]) -> List[int]:
+    if dim is None: return []
+    for d in dim:
+        assert not (d < -len(self) or d >= len(self))
+    return upstream_shape_functions.sum_mean_dim(self, dim,False,0)
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1454]: Ensure proper formatting in `sum_mean_dim(self, dim,False,0)`

Comment (vivekkhandelwal1) [reply projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1454]: Also, add what does `False` and `0` corresponds to in comment like: `/*keepdim=*/`

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 4]
```diff
@@ -2496,3 +2496,54 @@ def TraceUnsignedIntModule_basic(module, tu: TestUtils):
...
+
+
+class CountNonzeroDimIntListModuleBool(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.dim = [-2, -0, -1]
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py:2539]: Keep one dim so that this case will also be covered.

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 5]
```diff
@@ -6033,6 +6033,35 @@ LogicalResult AtenRot90Op::verify() {
...
+  auto selfType = cast<BaseTensorType>(getSelf().getType());
+
+  if (!selfType.hasDtype() || !selfType.hasSizes())
+    return success();
+
+  SmallVector<int64_t> dims;
+  if (!matchPattern(getDim(), m_TorchListOfConstantInts(dims)))
+    return emitOpError("expected dim to be constructed from list construct");
+  if (!isa<Torch::IntType>(getDim().getType().getContainedType()))
+    return emitOpError("list elements must be int type");
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/IR/TorchOps.cpp:6051]: This separate check is not required. Since, the above patern matching covers that.

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 6]
```diff
@@ -6033,6 +6033,35 @@ LogicalResult AtenRot90Op::verify() {
...
+
+  int64_t selfRank;
+  selfRank = selfType.getSizes().size();
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/IR/TorchOps.cpp:6054]: Keep them in single line.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 7]
```diff
@@ -6661,6 +6661,48 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+        !isa<Torch::IntType>(
+            (cast<Torch::ListType>(dimList.getType())).getContainedType())) {
+      return rewriter.notifyMatchFailure(
+          op, "expected `dim` to be constructed from list");
+    }
+    SmallVector<Value> dimListElements;
+    if (!getListConstructElements(dimList, dimListElements)) {
+      return rewriter.notifyMatchFailure(
+          op, "expected `dim` to be constructed from list");
+    }
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6683]: Instead of doing all this, you can just add a check for none and if found non-none then just use: `m_TorchListOfConstantInts`.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 8]
```diff
@@ -6661,6 +6661,48 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+    auto C_0 =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(0));
+    auto Ne_0 = rewriter.create<AtenNeScalarOp>(loc, boolTenTy, self, C_0);
+    auto none = rewriter.create<ConstantNoneOp>(loc);
+    auto C_False = rewriter.create<ConstantBoolOp>(loc, false);
+
+    auto newOp = rewriter.create<AtenSumDimIntListOp>(
+        loc, op.getResult().getType(), Ne_0, dimList, C_False, none);
+
+    rewriter.replaceOp(op, newOp);
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6699]: Use meaningful and relevant variable names as asked in your other PR, and avoid using `auto`.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 9]
```diff
@@ -6661,6 +6661,48 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+    : public OpRewritePattern<AtenCountNonzeroDimIntListOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenCountNonzeroDimIntListOp op,
+                                PatternRewriter &rewriter) const override {
+    Value dimList = op.getDim();
+    if (isa<Torch::NoneType>(dimList.getType())) {
+      return rewriter.notifyMatchFailure(
+          op, "expected `dim` to be constructed from list");
+    }
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6676]: Nit: Braces are not required.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 10]
```diff
@@ -6661,6 +6661,48 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+    Value nonZeroMask =
+        rewriter.create<AtenNeScalarOp>(loc, inpBoolTy, self, cstZero);
+    Value none = rewriter.create<ConstantNoneOp>(loc);
+    Value cstFalse = rewriter.create<ConstantBoolOp>(loc, false);
+
+    Value newOp = rewriter.create<AtenSumDimIntListOp>(
+        loc, op.getResult().getType(), nonZeroMask, dimList, cstFalse, none);
+
+    rewriter.replaceOp(op, newOp);
+
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6700]: Delete this extra line

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 11]
```diff
@@ -6661,6 +6661,48 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(0));
+    Value nonZeroMask =
+        rewriter.create<AtenNeScalarOp>(loc, inpBoolTy, self, cstZero);
+    Value none = rewriter.create<ConstantNoneOp>(loc);
+    Value cstFalse = rewriter.create<ConstantBoolOp>(loc, false);
+
+    Value newOp = rewriter.create<AtenSumDimIntListOp>(
+        loc, op.getResult().getType(), nonZeroMask, dimList, cstFalse, none);
+
+    rewriter.replaceOp(op, newOp);
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6699]: Instead, you can do: `rewriter.replaceOpWithNewOp<>()....`

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

**Comment by um-urosmarkovic** [comment]
@vivekkhandelwal1 I've addressed your comments. Let me know if it looks good. Thanks.

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

**Comment by um-urosmarkovic** [comment]
Thanks a lot, @vivekkhandelwal1 ! Could you please help merge this when you have a moment? I don’t have permissions. Much appreciated!

**Comment by vivekkhandelwal1** [comment]
> Thanks a lot, @vivekkhandelwal1 ! Could you please help merge this when you have a moment? I don’t have permissions. Much appreciated!

Hi @um-urosmarkovic, you need to resolve these conflicts.

**Comment by um-urosmarkovic** [comment]
Hi @vivekkhandelwal1 , I have addressed all conflicts. Can you check it when you get a chance? Thanks in advance.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 12]
```diff
@@ -12063,7 +12099,6 @@ class DecomposeComplexOpsPass
     addPatternIfTargetOpIsIllegal<DecomposeAtenVarCorrectionOp>(patterns);
     addPatternIfTargetOpIsIllegal<DecomposeAtenStdDimOp>(patterns);
     addPatternIfTargetOpIsIllegal<DecomposeAtenRot90Op>(patterns);
-    addPatternIfTargetOpIsIllegal<DecomposeAtenCountNonzeroOp>(patterns);
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12066]: Why is this removed?

Comment (um-urosmarkovic) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12066]: Apologies, it is resolved now. Thanks!

**Review by vivekkhandelwal1 — APPROVED** [review]
